### PR TITLE
Add support for async/await in the PCL project

### PR DESCRIPTION
**Pull request by flagbug**
This PR adds support for the asynchronous methods in the PCL project by removing support for Windows Phone Silverlight, which is basically dead anyway.

### CODE REVIEW

File: build/Stripe.net.nuspec [hunk 1]
```diff
@@ -29,7 +26,7 @@
     <file src="net45\Stripe.net.pdb" target="lib\net45" />
     <file src="portable\Stripe.net.dll" target="lib\dotnet" />
     <file src="portable\Stripe.net.pdb" target="lib\dotnet" />
-    <file src="portable\Stripe.net.dll" target="lib\portable-net45+wp8+wpa81+win8+MonoAndroid+MonoTouch+Xamarin.iOS+Xamarin.Mac" />
-    <file src="portable\Stripe.net.pdb" target="lib\portable-net45+wp8+wpa81+win8+MonoAndroid+MonoTouch+Xamarin.iOS+Xamarin.Mac" />
+    <file src="portable\Stripe.net.dll" target="lib\portable-net45+wp8+wpa81" />
```
Comment (jaymedavis) [inline build/Stripe.net.nuspec]: Why did you remove Xamarin?

Comment (flagbug) [inline build/Stripe.net.nuspec]: The Xamarin target is automatically infered by NuGet when specifying `portable-net45+wp8+wpa81` should have said that in the PR description, sorry!

**Comment by jaymedavis** [comment]
I don't think we can remove the flags around async. Mono does not support it to my knowledge.

**Comment by flagbug** [comment]
Mono support async/await just fine, since a few years even, if I remember correctly

**Comment by flagbug** [comment]
Source: I've written a few Xamarin apps and libraries :smile:

**Comment by xt0rted** [comment]
Mono's compatibility can be found here http://www.mono-project.com/docs/about-mono/compatibility/

**Comment by jaymedavis** [comment]
Interesting. Thanks.

**Comment by jaymedavis** [comment]
While I have your attention (and Xamarin experience) - I have another ticket for submitting this app to the Xamarin store. I've done the work already packaging it up, but another friend stated NuGet is enough. Would this being on the Xamarin store be a lot easier?

File: src/Stripe.Portable/Stripe.Portable.csproj [hunk 2]
```diff
@@ -438,33 +438,13 @@
...
   </ItemGroup>
   <ItemGroup>
-    <None Include="packages.config" />
-  </ItemGroup>
-  <ItemGroup>
     <Reference Include="Newtonsoft.Json, Version=8.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
       <HintPath>..\packages\Newtonsoft.Json.8.0.3\lib\portable-net40+sl5+wp80+win8+wpa81\Newtonsoft.Json.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="System.Net.Http, Version=1.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
```
Comment (jaymedavis) [inline src/Stripe.Portable/Stripe.Portable.csproj]: Could you comment on removing all of these libraries? Due to my lack of example apps and testing, I'm a little nervous to remove them. I'm very pro this PR though, I love deleting things. :facepunch:

Comment (flagbug) [inline src/Stripe.Portable/Stripe.Portable.csproj]: The references were only required for the Windows 8 Silverlight support. All modern platforms have these things built-in

**Comment by flagbug** [comment]
@jaymedavis I'm not a huge fan of the Xamarin Store, in my opinion it's just a shitty version of NuGet that has no dependency management whatsoever

**Comment by jaymedavis** [comment]
Cool, thanks. I have some user agent data I need to start collecting soon so I can track things like silverlight usage (i suspect you are right). This is something I will work on with Stripe soon, so I can collect the data from them. 

As for this PR, I'd like to make these changes sooner rather than later, though - so I want to give a chance for @ry8806, @advancedrei, @chaitanyakale, and @chickenham to comment. They all helped a lot with the PCL addition to Stripe.net. 

Thanks @flagbug.

**Comment by ry8806** [comment]
think this is a great idea, bringing the library forward sooner rather than later (and avoiding problems later on down the line).

I'm in favour of removing support for [Windows Phone Silverlight](https://en.wikipedia.org/wiki/Windows_Phone_7) - the wiki page says MS ended support for it Oct 2014 - so I definitely think we're safe removing support - can't see anyone creating a new WP7 app in the near future :stuck_out_tongue:

**Comment by flagbug** [comment]
What I also want to do is migrating the library to project.json and [.NET Platform Standard](https://github.com/dotnet/corefx/blob/ec2685715b01d12f16b08d0dfa326649b12db8ec/Documentation/architecture/net-platform-standard.md)  if that's OK with everyone.

**Comment by chaitanyakale** [comment]
Let it go! UWP, Android, iOS with PCL that's about it.

**Comment by jaymedavis** [comment]
I have moved the migrating conversation to #544, to keep this one focused on just the PCL change. I know they are similar, but I'd rather migrate in the work separately since this one is relatively easy. I've left some comments there.

**Comment by jaymedavis** [comment]
Couple questions if you want to knock it out here, I left one on #544 (I closed because I realized YOU wanted to handle this, sorry for the confusion)k and another you may be familiar with - can we move the entire library to one project? This sounds like a lot of work, but it's fairly easy to modify the project files manually to move the references. I think I could knock that out. Let me know if you'd like me to start a branch @flagbug, and we can pair up on that.

**Comment by advancedrei** [comment]
I think you should do a Major version release that pulls Stripe out of all the service names and moves back to one project where everything is "portable", using .NET Standard 1.3 or greater. Would be a great opportunity to get rid of some cruft, go full Async, and make some architectural adjustments.

HTH!

**Comment by jaymedavis** [comment]
Keep this in mind as well @flagbug. https://stripe.com/blog/upgrading-tls

**Comment by jaymedavis** [comment]
Let's definitely not remove non async right now. Killing the non async versions kills 700+ tests. :-1:

I'm all for it as a goal, but it's too much work to bite off for this.

**Comment by flagbug** [comment]
> Let's definitely not remove non async right now. Killing the non async versions kills 622 tests. :-1:

I mean this is just an idea, but we could make all the synchronous methods `internal` and expose them to the test project via `InternalsVisibleTo`, so the tests keep running but the consumers only see the `async` methods.

**Comment by flagbug** [comment]
Ok, so as I said in #544, it would be great to get this PR merged in, and as soon as I have the time, I'll create a new PR moving this project to `project.json` do all the other maintenance work :+1:

**Comment by jaymedavis** [comment]
I just don't want to force async unless I know it's what the users of the library want. I'm working on gathering more info about the platforms used with Stripe.net so these kind of decisions are easier.

And Thanks - I'll work on getting this in. :+1:

File: src/Stripe/Infrastructure/Requestor.cs [hunk 3]
```diff
@@ -150,18 +148,15 @@ private static string ExecuteRequest(HttpRequestMessage requestMessage)
...
         private static async Task<string> ExecuteRequestAsync(HttpRequestMessage requestMessage)
         {
             var response = await HttpClient.SendAsync(requestMessage);
 
             if (response.IsSuccessStatusCode)
                 return await response.Content.ReadAsStringAsync();
 
-            // this is not async
-            throw BuildStripeException(response.StatusCode, requestMessage.RequestUri.AbsoluteUri, response.Content.ReadAsStringAsync().Result);
+            throw BuildStripeException(response.StatusCode, requestMessage.RequestUri.AbsoluteUri, await response.Content.ReadAsStringAsync());
```
Comment (jaymedavis) [inline src/Stripe/Infrastructure/Requestor.cs]: How very kind of you to review this :)... I am fuzzy on the code here, because I am using .Result in my code. I don't think it matters, however, because it's an exceptional case.
